libtfhe/core/arithmetic/polynomial_torus.cpp

```diff
@@ -0,0 +1,338 @@
+{
+    const int32_t N = params->N;
+    TORUS *x = result->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        x[i] = RandomGenTorus<TORUS>::uniform();
+}
+
+// TorusPolynomial + p*TorusPolynomial
+template<typename TORUS>
+void TorusPolynomial<TORUS>::AddMulZ(
+    TorusPolynomial<TORUS> *result,
+    const TorusPolynomial<TORUS> *poly1,
+    const INT_TYPE *p,
+    const TorusPolynomial<TORUS> *poly2,
+    const PolynomialParams<TORUS> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    const int32_t N = params->N;
+    assert(result != poly1); //if it fails here, please use AddMulZTo
+    TORUS *r = result->coefs;
+    const TORUS *a = poly1->coefs;
+    const TORUS *b = poly2->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        r[i] = a[i] + *p * b[i];
+}
+
+// TorusPolynomial += p*TorusPolynomial
+template<typename TORUS>
+void TorusPolynomial<TORUS>::AddMulZTo(
+    TorusPolynomial<TORUS> *result,
+    const INT_TYPE *p,
+    const TorusPolynomial<TORUS> *poly2,
+    const PolynomialParams<TORUS> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    const int32_t N = params->N;
+    TORUS *r = result->coefs;
+    const TORUS *b = poly2->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        r[i] += *p * b[i];
+}
+
+// TorusPolynomial - p*TorusPolynomial
+template<typename TORUS>
+void TorusPolynomial<TORUS>::SubMulZ(
+    TorusPolynomial<TORUS> *result,
+    const TorusPolynomial<TORUS> *poly1,
+    const INT_TYPE *p,
+    const TorusPolynomial<TORUS> *poly2,
+    const PolynomialParams<TORUS> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    const int32_t N = params->N;
+    assert(result != poly1); //if it fails here, please use SubMulZTo
+    TORUS *r = result->coefs;
+    const TORUS *a = poly1->coefs;
+    const TORUS *b = poly2->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        r[i] = a[i] - *p * b[i];
+}
+
+// TorusPolynomial -= p*TorusPolynomial
+template<typename TORUS>
+void TorusPolynomial<TORUS>::SubMulZTo(
+    TorusPolynomial<TORUS> *result,
+    const INT_TYPE *p,
+    const TorusPolynomial<TORUS> *poly2,
+    const PolynomialParams<TORUS> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    const int32_t N = params->N;
+    TORUS *r = result->coefs;
+    const TORUS *b = poly2->coefs;
+
+    for (int32_t i = 0; i < N; ++i)
+        r[i] -= *p * b[i];
+}
+
+
+// Infinity norm of the distance between two TorusPolynomial
+template<typename TORUS>
+double TorusPolynomial<TORUS>::NormInftyDist(
+    const TorusPolynomial<TORUS> *poly1,
+    const TorusPolynomial<TORUS> *poly2,
+    const PolynomialParams<TORUS> *params,
+    TfheThreadContext *context,
+    Allocator alloc)
+{
+    const int32_t N = params->N;
+    double norm = 0;
+
+    // Max between the coefficients of abs(poly1-poly2)
+    for (int32_t i = 0; i < N; ++i) {
+        double r = abs(TorusUtils<TORUS>::to_double(poly1->coefs[i] - poly2->coefs[i], params->zmodule_params));
```

Comment:
This template assumes that Torus is signed. 

It is the case here if we only use int32_t and int64_t, but either we need to clarify this in the documentation, or alternatively,  it would be better to define a ```double TorusUtils<TORUS>::normInftyDist``` between two torus.

Reply:
I agree. In the case of BigTorus will a double return type have sufficient precision ?

Reply:
yes, this function will be used only for tests, and the exponent of the double will have enough precision